### ISSUE TITLE
Services running in Docker become unreachable after minutes or hours

### ISSUE THREAD

**Issue by ruverav** [docker/for-mac#3526]
- [x] I have tried with the latest version of my channel (Stable or Edge)
  - [x] I have uploaded Diagnostics
  - Diagnostics ID: F13CDAF6-3BE1-4DBE-A233-C871C7D4F745/20190214075450

### Expected behavior

Docker should run flawlessly for at least  days or months.

### Actual behavior

Docker stops working or becomes unreachable without any specific reason after minutes or hours of running without issues. The last time I checked out, a month ago or so, I had a too many TCP connections type of error. It happens to me even with older versions of Docker. I' m so fed of this problem I'm actually thinking about installing Linux or Windows and using Docker there.

### Information

  - macOS Version: High Sierra 10.13.8

### Diagnostic logs

```
Docker for Mac: version... 2.0.0.2

time="2019-02-14T08:54:51+01:00" level=warning msg="[/usr/local/bin/kubectl.docker --namespace docker get pods -o go-template={{range .items}}{{.metadata.name}}\n{{end}}]: exit status 1" type=kubernetes
time="2019-02-14T08:54:51+01:00" level=warning msg="The connection to the server localhost:8080 was refused - did you specify the right host or port?\n" type=kubernetes
time="2019-02-14T08:54:51+01:00" level=warning msg="[/usr/local/bin/kubectl.docker --namespace kube-system get pods -o go-template={{range .items}}{{.metadata.name}}\n{{end}}]: exit status 1" type=kubernetes
time="2019-02-14T08:54:51+01:00" level=warning msg="The connection to the server localhost:8080 was refused - did you specify the right host or port?\n" type=kubernetes
time="2019-02-14T08:54:51+01:00" level=info msg="Start gathering list of 69 elements (timeout=10m0s)" type=list
time="2019-02-14T08:54:51+01:00" level=info msg="Executing: [sw_vers -productVersion]" type=cmd
time="2019-02-14T08:54:51+01:00" level=info msg="Executing: [sw_vers -buildVersion]" type=cmd
time="2019-02-14T08:54:51+01:00" level=info msg="Executing: [uname -r]" type=cmd
time="2019-02-14T08:54:51+01:00" level=info msg="Executing: [sysctl kern.boottime]" type=cmd
time="2019-02-14T08:54:51+01:00" level=info msg="Executing: [sysctl hw.memsize]" type=cmd
time="2019-02-14T08:54:51+01:00" level=info msg="Executing: [vm_stat]" type=cmd
time="2019-02-14T08:54:51+01:00" level=info msg="Executing: [sysctl machdep.cpu]" type=cmd
time="2019-02-14T08:54:52+01:00" level=info msg="Executing: [uname -a]" type=cmd
time="2019-02-14T08:54:52+01:00" level=info msg="Executing: [ps uax]" type=cmd
time="2019-02-14T08:54:52+01:00" level=info msg="Executing: [dmesg]" type=cmd
time="2019-02-14T08:54:52+01:00" level=warning msg="[dmesg] exit status 1" type=cmd
time="2019-02-14T08:54:52+01:00" level=warning msg="[dmesg]: Unable to obtain kernel buffer: Operation not permitted\nusage: sudo dmesg\n" type=cmd
time="2019-02-14T08:54:52+01:00" level=info msg="Executing: [id]" type=cmd
time="2019-02-14T08:54:52+01:00" level=info msg="Executing: [sysctl -a]" type=cmd
time="2019-02-14T08:54:52+01:00" level=info msg="Executing: [mount]" type=cmd
time="2019-02-14T08:54:52+01:00" level=info msg="Executing: [df -h]" type=cmd
time="2019-02-14T08:54:52+01:00" level=info msg="Executing: [launchctl print-disabled system]" type=cmd
time="2019-02-14T08:54:52+01:00" level=info msg="Executing: [system_profiler SPHardwareDataType]" type=cmd
time="2019-02-14T08:54:52+01:00" level=info msg="Executing: [defaults read /Library/Preferences/com.apple.alf.plist]" type=cmd
time="2019-02-14T08:54:52+01:00" level=info msg="Executing: [ifconfig -a]" type=cmd
time="2019-02-14T08:54:52+01:00" level=info msg="Executing: [netstat -ln]" type=cmd
time="2019-02-14T08:54:52+01:00" level=info msg="Executing: [netstat -mm]" type=cmd
time="2019-02-14T08:54:52+01:00" level=info msg="Executing: [netstat -nr]" type=cmd
time="2019-02-14T08:54:52+01:00" level=info msg="Gathering: /etc/resolv.conf" type=file
time="2019-02-14T08:54:52+01:00" level=info msg="Executing: [scutil --dns]" type=cmd
time="2019-02-14T08:54:52+01:00" level=info msg="Executing: [scutil --nc list]" type=cmd
time="2019-02-14T08:54:53+01:00" level=info msg="Executing: [scutil --proxy list]" type=cmd
time="2019-02-14T08:54:53+01:00" level=info msg="Executing: [system_profiler SPFirewallDataType]" type=cmd
time="2019-02-14T08:54:53+01:00" level=info msg="HTTP GET: localhost /v1.24/version" type=http
time="2019-02-14T08:54:53+01:00" level=info msg="HTTP GET: localhost /v1.24/info" type=http
time="2019-02-14T08:54:53+01:00" level=info msg="HTTP GET: localhost /v1.24/containers/json?all=1" type=http
time="2019-02-14T08:54:53+01:00" level=info msg="HTTP GET: localhost /v1.24/nodes" type=http
time="2019-02-14T08:54:53+01:00" level=info msg="HTTP GET: localhost /v1.24/services" type=http
time="2019-02-14T08:54:53+01:00" level=info msg="HTTP GET: localhost /v1.24/networks" type=http
time="2019-02-14T08:54:53+01:00" level=info msg="Executing: [/bin/sh -c docker network inspect $(docker network ls -q)]" type=cmd
time="2019-02-14T08:54:54+01:00" level=warning msg="kubernetes context cannot be reached. Ignoring gatherer:describepods" type=kubectl
time="2019-02-14T08:54:54+01:00" level=warning msg="kubernetes context cannot be reached. Ignoring gatherer:describedaemonset" type=kubectl
time="2019-02-14T08:54:54+01:00" level=warning msg="kubernetes context cannot be reached. Ignoring gatherer:describereplicaset" type=kubectl
time="2019-02-14T08:54:54+01:00" level=warning msg="kubernetes context cannot be reached. Ignoring gatherer:describedeployments" type=kubectl
time="2019-02-14T08:54:54+01:00" level=warning msg="kubernetes context cannot be reached. Ignoring gatherer:describepods" type=kubectl
time="2019-02-14T08:54:54+01:00" level=warning msg="kubernetes context cannot be reached. Ignoring gatherer:describedaemonset" type=kubectl
time="2019-02-14T08:54:54+01:00" level=warning msg="kubernetes context cannot be reached. Ignoring gatherer:describereplicaset" type=kubectl
time="2019-02-14T08:54:54+01:00" level=warning msg="kubernetes context cannot be reached. Ignoring gatherer:describedeployments" type=kubectl
time="2019-02-14T08:54:54+01:00" level=info msg="Executing: [/bin/sh -c /usr/bin/log show --debug --info --style syslog --last 1d --predicate 'process matches \".*[Dd]ocker.*\" || (process in {\"kernel\", \"launchservicesd\", \"taskgated-helper\"} && eventMessage contains[c] \"docker\")']" type=cmd
time="2019-02-14T08:55:05+01:00" level=info msg="Executing: [syslog -F $((Time)(JZ)) $Host $(Sender)[$(Facility)][$(PID)] <$((Level)(str))>: $Message -o -k Sender Seq docker -o -k Sender Seq Docker -o -k Message Seq docker -o -k Message Seq Docker]" type=cmd
time="2019-02-14T08:55:05+01:00" level=warning msg="[syslog -F $((Time)(JZ)) $Host $(Sender)[$(Facility)][$(PID)] <$((Level)(str))>: $Message -o -k Sender Seq docker -o -k Sender Seq Docker -o -k Message Seq docker -o -k Message Seq Docker]: NOTE:  Most system logs have moved to a new logging system.  See log(1) for more information.\n" type=cmd
time="2019-02-14T08:55:05+01:00" level=info msg="Executing: [/bin/sh -c lsof -n -P -p $(pgrep -o com.docker.driver.amd64-linux)]" type=cmd
time="2019-02-14T08:55:05+01:00" level=info msg="Executing: [/bin/sh -c lsof -n -P -p $(pgrep -o com.docker.hyperkit)]" type=cmd
time="2019-02-14T08:55:05+01:00" level=info msg="Executing: [/bin/sh -c lsof -n -P -p $(pgrep -o com.docker.osxfs)]" type=cmd
time="2019-02-14T08:55:05+01:00" level=info msg="Executing: [/bin/sh -c lsof -n -P -p $(pgrep -o com.docker.supervisor)]" type=cmd
time="2019-02-14T08:55:05+01:00" level=info msg="Executing: [/bin/sh -c lsof -n -P -p $(pgrep -o com.docker.vpnkit)]" type=cmd
time="2019-02-14T08:55:05+01:00" level=info msg="Executing: [/bin/sh -c lsof -n -P -p $(pgrep -o com.docker.vmnetd)]" type=cmd
time="2019-02-14T08:55:05+01:00" level=warning msg="[/bin/sh -c lsof -n -P -p $(pgrep -o com.docker.vmnetd)] exit status 1" type=cmd
time="2019-02-14T08:55:05+01:00" level=warning msg="kubernetes context cannot be reached. Ignoring gatherer:version" type=kubectl
time="2019-02-14T08:55:06+01:00" level=warning msg="kubernetes context cannot be reached. Ignoring gatherer:configget-contexts" type=kubectl
time="2019-02-14T08:55:06+01:00" level=warning msg="kubernetes context cannot be reached. Ignoring gatherer:configcurrent-context" type=kubectl
time="2019-02-14T08:55:06+01:00" level=warning msg="kubernetes context cannot be reached. Ignoring gatherer:getnodes" type=kubectl
time="2019-02-14T08:55:06+01:00" level=warning msg="kubernetes context cannot be reached. Ignoring gatherer:getnamespaces" type=kubectl
time="2019-02-14T08:55:06+01:00" level=warning msg="kubernetes context cannot be reached. Ignoring gatherer:describenodes" type=kubectl
time="2019-02-14T08:55:06+01:00" level=warning msg="kubernetes context cannot be reached. Ignoring gatherer:describepods" type=kubectl
time="2019-02-14T08:55:06+01:00" level=warning msg="kubernetes context cannot be reached. Ignoring gatherer:cluster-infodump" type=kubectl
time="2019-02-14T08:55:06+01:00" level=warning msg="kubernetes context cannot be reached. Ignoring gatherer:describeservices" type=kubectl
time="2019-02-14T08:55:06+01:00" level=warning msg="kubernetes context cannot be reached. Ignoring gatherer:describestorageclass,pv,pvc" type=kubectl
time="2019-02-14T08:55:06+01:00" level=info msg="Executing: [/Applications/Docker.app/Contents/MacOS/qcow-tool check /Users/Ruben/Library/Containers/com.docker.docker/Data/vms/0/Docker.qcow2]" type=cmd
time="2019-02-14T08:55:07+01:00" level=warning msg="[/Applications/Docker.app/Contents/MacOS/qcow-tool check /Users/Ruben/Library/Containers/com.docker.docker/Data/vms/0/Docker.qcow2]: qcow-tool: [INFO] image has 501 free sectors and 138270 used sectors\n" type=cmd
time="2019-02-14T08:55:07+01:00" level=info msg="Gathering: /Users/Ruben/Library/Group Containers/group.com.docker/DockerAppStderr.txt" type=file
time="2019-02-14T08:55:07+01:00" level=info msg="Gathering: /Users/Ruben/Library/Group Containers/group.com.docker/already-enabled-features.json" type=file
time="2019-02-14T08:55:07+01:00" level=info msg="Gathering: /Users/Ruben/Library/Group Containers/group.com.docker/dhcp.json" type=file
time="2019-02-14T08:55:07+01:00" level=info msg="Gathering: /Users/Ruben/Library/Group Containers/group.com.docker/features.json" type=file
time="2019-02-14T08:55:07+01:00" level=info msg="Gathering: /Users/Ruben/Library/Group Containers/group.com.docker/http_proxy.json" type=file
time="2019-02-14T08:55:07+01:00" level=info msg="Gathering: /Users/Ruben/Library/Group Containers/group.com.docker/last-start-linux-daemon.json" type=file
time="2019-02-14T08:55:07+01:00" level=info msg="Gathering: /Users/Ruben/Library/Group Containers/group.com.docker/settings.json" type=file
time="2019-02-14T08:55:07+01:00" level=warning msg="/Users/Ruben/Library/Preferences/group.com.docker.pliststat /Users/Ruben/Library/Preferences/group.com.docker.plist: no such file or directory" type=file
time="2019-02-14T08:55:07+01:00" level=info msg="Gathering: /Users/Ruben/.docker/config.json" type=redacted-json
time="2019-02-14T08:55:07+01:00" level=info msg="Gathering: /Users/Ruben/.docker/daemon.json" type=redacted-json
time="2019-02-14T08:55:07+01:00" level=info msg="Getting Linux Diagnostics via a virtio socket: hyperkit:vms/0:62374" type=virtsock
time="2019-02-14T08:55:41+01:00" level=info msg="All 69 gatherers finished" type=list


```

### Steps to reproduce the behavior

  1. ... Run a nginx proxy with several services, sonarr, radarr, qbittorrent, jackett, etc..
  2. ... Wait for it.

**Comment by BEisem** [comment]
I'm seeing this as well, it happens either more frequently or less frequently, depending on which containers I am running. If I run varken, which polls every 30 seconds, it will happen within a matter of hours.

Here is my Diagnostic, to add to the report: 6C1F7F8C-1CFC-4C8F-83F8-C6379EB35432/20190216193719

Also, I found in another thread the following command:

lsof -p $(pgrep vpnkit) | wc -l

I'll be honest I don't know exactly what it does, but it outputs a number. I believe this is the number of active connections. When I first start Docker, this number is relatively low (in the 20s). As Docker runs, the number continues to climb until it reaches about 2000. At that point, Docker becomes nonresponsive and I have to restart in order to be able to communicate with it.

**Comment by BEisem** [comment]
By the way I posted a similar issue a few days ago, but have not had it assigned yet. My issue can be found [https://github.com/docker/for-mac/issues/3518](here).

**Comment by pawoolley** [comment]
Also happening for me and my colleagues on a regular (e.g. daily) basis too.  Throwing my diagnostic info in to the mix:  8D6AF156-E1F6-48BE-825D-9E5E25A9076C/20190305133607

Docker: 2.0.0.3
Mac OS: 10.14.3

Nothing besides a docker restart fixes it.  🤞for a fix soon.

**Comment by microtherion** [comment]
Experiencing the same thing on  a regular bases. Diagnostic ID F4D86769-BF59-48DA-8B48-0E7D6FDFA353/20190313230726

**Comment by vseibt** [comment]
Same problem, Diagnostic ID 609D0BF3-5FC6-44DA-9567-DFBB0781DE83/20190430072857.

I'm not sure if this is really a problem with the count of connections, `lsof -p $(pgrep vpnkit) | wc -l` for me returns 154.

**Comment by ruverav** [comment]
If it does help anybody I have been running [this LaunchDaemon I created](https://github.com/ruverav/docker-mac-restarter) for MacOS on my Mac Mini for several months and I haven't had to manually restart Docker since then, since it doest restart it automatically for you. The more common scenario is to check connectivity against a nginx proxy running on localhost:80, or at least it's how I use it.

Also be sure to check that container you are pointing to restarts automatically (flag "restart" must be set to "always") so it starts again when Docker reboots and the restarter daemon isn't triggered again without actually being needed.

**Comment by dmuth** [comment]
This sounds similar to [the issue I just reported](https://github.com/docker/for-mac/issues/3674).  Though I had not thought to try an `lsof` and will look into that...

That said, I wrote an app to test TCP connectivity to Docker, in case anyone here finds it useful: https://github.com/dmuth/docker-health-check  It will test nginx once per second, and after too many timeouts, will attempt to restart nginx until success.  (In other words, if you restart Docker, the tests will resume)

Hope this helps.
